packages/react-dom-bindings/src/client/ReactFiberConfigDOM.js

```diff
@@ -1534,6 +1534,206 @@ export function subscribeToGestureDirection(
+    optionsOrUseCapture?: EventListenerOptionsOrUseCapture,
+  ): void,
+  removeEventListener(
+    type: string,
+    listener: EventListener,
+    optionsOrUseCapture?: EventListenerOptionsOrUseCapture,
+  ): void,
+  focus(): void,
+};
+
+function FragmentInstancePseudoElement(
+  this: FragmentInstance,
+  fragmentFiber: Fiber,
+) {
+  this._fragmentFiber = fragmentFiber;
+}
+// $FlowFixMe[prop-missing]
+FragmentInstancePseudoElement.prototype.addEventListener = function (
+  this: FragmentInstance,
+  type: string,
+  listener: EventListener,
+  optionsOrUseCapture?: EventListenerOptionsOrUseCapture,
+): void {
+  if (!this._eventListeners) {
+    this._eventListeners = [];
+  }
+
+  const listeners = this._eventListeners;
+  // Element.addEventListener will only apply uniquely new event listeners by default. Since we
+  // need to collect the listeners to apply to appended children, we track them ourselves and use
+  // custom equality check for the options.
+  const isNewEventListener =
+    indexOfEventListener(listeners, {
+      type,
+      listener,
+      optionsOrUseCapture,
+    }) === -1;
+  if (isNewEventListener) {
+    listeners.push({type, listener, optionsOrUseCapture});
+    const children = getFragmentInstanceChildren(this);
+    children.forEach(child => {
+      child.addEventListener(type, listener, optionsOrUseCapture);
+    });
+  }
+  this._eventListeners = listeners;
+};
+// $FlowFixMe[prop-missing]
+FragmentInstancePseudoElement.prototype.removeEventListener = function (
+  this: FragmentInstance,
+  type: string,
+  listener: EventListener,
+  optionsOrUseCapture?: EventListenerOptionsOrUseCapture,
+): void {
+  const listeners = this._eventListeners;
+  if (listeners === undefined || listeners.length === 0) {
+    return;
+  }
+
+  const children = getFragmentInstanceChildren(this);
+  children.forEach(child => {
+    child.removeEventListener(type, listener, optionsOrUseCapture);
+  });
+  const index = indexOfEventListener(listeners, {
+    type,
+    listener,
+    optionsOrUseCapture,
+  });
+  this._eventListeners = listeners.splice(index, 1);
+};
+// $FlowFixMe[prop-missing]
+FragmentInstancePseudoElement.prototype.focus = function (
+  this: FragmentInstance,
+) {
+  const children = getFragmentInstanceChildren(this);
+  const iterator = children.values();
+  let child = iterator.next();
+  while (!child.done) {
+    if (setFocusIfFocusable(child.value)) {
+      break;
+    }
+    child = iterator.next();
+  }
+};
+
+function getFragmentInstanceChildren(
+  fragmentInstance: FragmentInstance,
+): Set<Element> {
+  const children = new Set<Element>();
+  recursivelyGetFragmentInstanceChildren(
+    fragmentInstance._fragmentFiber.child,
+    children,
+  );
+  return children;
+}
+
+function recursivelyGetFragmentInstanceChildren(
+  child: Fiber | null,
+  childElements: Set<Element>,
+): void {
+  if (child === null) {
+    return;
+  }
+
+  if (child.tag === HostComponent) {
```

Comment:
This probably need to consider some other child tags as well. For example inside an `<Activity>` that's currently hidden we shouldn't discover any elements. (If we did then they would be duplicate when it gets revealed since they're added again.) So we need to bail out from searching inside Activity here.

Reply:
Added a boundary based on `OffscreenComponent`. Another that could be relevant is `HostPortal`, though everything in portals like events and context follow the React tree, so it seems like for consistency we should include any host nodes within portals here as well.